### ISSUE TITLE
Browser elements should respond to click events

### ISSUE THREAD

**Issue by singingwolfboy** [searchkit/searchkit#416]
The [`<FastClick>`](https://github.com/singingwolfboy/searchkit/blob/master/src/core/react/FastClick.ts) component wraps every other component in the codebase that handles user interaction in the browser. Ironically, this component does not actually support `click` events! It supports `mousedown` and `touchstart`/`touchend`, but [`click` events are swallowed and ignored](https://github.com/singingwolfboy/searchkit/blob/546daad61fbc426f3fed5c5b2e1bdc5ed48810d6/src/core/react/FastClick.ts#L60-L62).

Not all users use a mouse or a touchscreen. If a user tries to navigate the website using a keyboard (or an assistive tech device that emulates a keyboard), the user will be totally unable to use Searchkit. (Keyboards can send `click` events by focusing on a clickable element like a `<button>`, and pressing spacebar.) In order to be accessible, Searchkit needs to support keyboard use.

I see two ways of addressing this issue: removing the `<FastClick>` component entirely, or making it support `click` events. I'm not sure which approach is better, but I'd like to start the discussion. Why does the `<FastClick>` component exist? Why is it specifically designed to swallow and ignore `click` events?

This is related to #359 (accessibility).

**Comment by joemcelroy** [comment]
FastClick was a component we built in order to speed up the time between the user clicking an action to seeing results. click event is slower than mousedown (we got obsessive over speed :) ). 

One way is for fastclick's click event to swallow only if there's an event that been dispatched by mousedown.

Im happy to hear any other suggestions.

**Comment by singingwolfboy** [comment]
How would FastClick be able to reliably determine if there was an event dispatched by `mousedown`, when deciding whether or not to swallow a `click` event? If your concern is about performance, I would expect that holding on to prior events is going to cause a performance penalty.

There's also a nice usability perk that comes from using the `click` event instead of the `mousedown` event -- the user can mousedown on an element, change their mind, move the mouse off the element, and then release the mouse in order to cancel the impending event. It's a little thing, but it's a bit disappointing that you can't do that with the way you've implemented FastClick.

I guess my question is, why is speed so important in this case? I agree that performance is an important metric, but not to the extent that it negates others like usability and accessibility.

**Comment by ssetem** [comment]
We could make it configurable so can switch off the optimization.

I think assuming performance is unimportant/important is an assumption, so we can make it configurable

**Comment by singingwolfboy** [comment]
The way I see it, the two options are:

* Squeeze out a tiny bit more performance, at the expense of making Searchkit inaccessible to keyboard users
* Make Searchkit accessible to everyone, at the expense of a tiny bit of performance

I believe that accessibility _matters_. I want to stress that performance matters as well! However, performance will not meaningfully suffer if we use the `click` event instead of `mousedown`. Accessibility _will_ meaningfully suffer if we use `mousedown` instead of `click`.

If you want to make this a configurable option, then we can do that, but even in that case, defaults matter. And the question that I'm trying to ask is, should Searchkit be accessible to everyone by default? I believe that it should. (And personally, I don't know why you would intentionally turn that off, but I'm willing to believe that someone might decide to make that trade-off.)

**Comment by ssetem** [comment]
yes, I agree so by giving the option to do both I feel we are doing our part.
We are also maintaining backwards compatibility, and we have only had 1 issue raised atm, but I agree accessibility is important :)

We also accept PR's :D

**Comment by joemcelroy** [comment]
There's a noticable difference between click and mousedown, so we took advantage of that. Didn't consider all the side effects when we implemented, we just wanted to make it fast. I don't think holding to prior would cause such a performance penalty to be an issue. 

We will probably go with switching off fastclick by default and allow users to opt in for this feature.

**Comment by singingwolfboy** [comment]
@ssetem I would be happy to send a pull request, _after_ we decide what the best course of action is here. I don't want to spend several hours writing code and modifying tests, only for us to decide that we need to do something different. Can you elaborate on your backwards compatibility concerns?

@joemcelroy That works. Can you describe a mechanism for how users might opt in for this feature? I can follow that example when making a pull request.

Alternatively, if one of you wants to make this change instead of waiting for my pull request, that's perfectly fine with me. 😄

**Comment by ssetem** [comment]
cool, will think about how its configured, might have to be a FastClick.STATIC.. or context based.
I think a static probably best in this case but not most functional

Backwards compat as in there are apps that depend on existing behaviour, if changing the defaults we need to be clear in change log etc

**Comment by GregoryPotdevin** [comment]
One solution could be to turn on FastClick by default on mobile, and off on desktop. It's mostly interesting for mobiles which have a bigger click delay (though you can remove it with `<meta name="viewport" content="width=device-width">`)

**Comment by ssetem** [comment]
yeh could be FastClick.configure(..options for both)

**Comment by MarttiR** [comment]
FastClick seems to be quite problematic on mobile: I have users finding search filters hard to use because they are "too sensitive". I agree, it's almost impossible to touch-scroll past the filters without accidentally activating something, which makes for a frustrating user experience.

I'm also seeing behavior where content inside the `Hits` component gets selected when user clicks on a `Pagination` button. This doesn't happen every time, though; it seems that sometimes a render manages to happen entirely after `mousedown` but before `mouseup`, in effect making the browser think that the user dragged the mouse to select text.

If the only reason for using FastClick is mobile performance, and if that performance can also be achieved using `<meta name="viewport" content="width=device-width">`, there should be a way to completely disable FastClick and/or hand in a plain click handler. The site I'm working on already uses that particular viewport setting anyway, as any modern sites should, so FastClick seems to be a solution to a non-problem.

In any case, using a different component for "mobile" and another for "desktop" sounds a bit magical and scary. Making this choice explicit by design and not automatic by default would probably be more predictable and thus more developer-friendly.

**Comment by ssetem** [comment]
we will look to making it configurable soon

**Comment by saracook** [comment]
Is there a way to turn off FastClick, even if it's hacky?

**Comment by ssetem** [comment]
right at beginning of your code, something alone these lines
```js
import {FastClick} from "searchkit"

FastClick.prototype.render = function(){
  return React.cloneElement(this.props.children, {
    onClick:this.props.handler
  })
}
```

**Comment by joemcelroy** [comment]
We are going to remove FastClick in the next version as not that much gain for alot of pain.

**Comment by joemcelroy** [comment]
Ash made a PR for this #464

**Comment by joemcelroy** [comment]
This has been merged and released on v2.1
